crates/latch_core/src/ecs/system_macros.rs

```diff
@@ -0,0 +1,181 @@
+/// });
+/// ```
+/// 
+/// # Parallel Iteration
+/// 
+/// For parallel iteration with rayon, use the `parallel` variant:
+/// 
+/// ```ignore
+/// use rayon::prelude::*;
+/// 
+/// for_each_entity!(world, [Position, Velocity], |
+///     (pos_curr, pos_next),
+///     (vel_curr, vel_next)
+/// | {
+///     // This closure will be called for each entity in parallel
+///     pos_next.x = pos_curr.x + vel_curr.x as i32;
+///     pos_next.y = pos_curr.y + vel_curr.y as i32;
+/// }, parallel);
+/// ```
+#[macro_export]
+macro_rules! for_each_entity {
+    // Sequential iteration: for_each_entity!(world, [C1, C2, ...], |args| { body })
+    ($world:expr, [$($Component:ty),+ $(,)?], |$(($curr:ident, $next:ident)),+ $(,)?| $body:expr) => {{
+        use $crate::ecs::Component;
+        
+        // Collect component IDs
+        let component_ids = [$(<$Component as Component>::ID),+];
+        
+        // Iterate over all archetypes with these components
+        $world.for_each_archetype_with_components(&component_ids, |storage| {
+            // Extract all read slices at once (current buffer)
+            let ($($curr),+) = $crate::columns!(storage, $($Component),+);
+            
+            // Extract all write slices at once (next buffer)
+            let ($($next),+) = $crate::columns_mut!(storage, $($Component),+);
+            
+            // Create iterator that zips all read and write slices together
+            let iter = $crate::for_each_entity!(@build_iter $($curr, $next),+);
+            
+            // Execute the user's closure for each entity
+            iter.for_each(|$crate::for_each_entity!(@pattern $($curr, $next),+)| $body);
+        });
+    }};
+    
+    // Parallel iteration: for_each_entity!(world, [C1, C2, ...], |args| { body }, parallel)
+    ($world:expr, [$($Component:ty),+ $(,)?], |$(($curr:ident, $next:ident)),+ $(,)?| $body:expr, parallel) => {{
+        use $crate::ecs::Component;
+        use rayon::prelude::*;
+        
+        // Collect component IDs
+        let component_ids = [$(<$Component as Component>::ID),+];
+        
+        // Iterate over all archetypes with these components
+        $world.for_each_archetype_with_components(&component_ids, |storage| {
+            // Extract all read slices at once (current buffer)
+            let ($($curr),+) = $crate::columns!(storage, $($Component),+);
+            
+            // Extract all write slices at once (next buffer)
+            let ($($next),+) = $crate::columns_mut!(storage, $($Component),+);
+            
+            // Create parallel iterator that zips all read and write slices together
+            let iter = $crate::for_each_entity!(@build_par_iter $($curr, $next),+);
+            
+            // Execute the user's closure for each entity in parallel
+            iter.for_each(|$crate::for_each_entity!(@pattern $($curr, $next),+)| $body);
+        });
+    }};
+    
+    // Helper: Build pattern for 1 component
```

Comment:
All of these helpers should not be necessary. For starters, they're verbose -- that's a lot of new lines of code. But to make matters worse they're inflexible: they limit us to 1, 2, 3, or 4 components. What if we want 7?

It should be possible to write macros which support arbitrary numbers of components. If our current public APIs make this challenging, the APIs can be updated.

This means we can either change the shape returned by world, archetype storage, or column methods... but it also means we can make a slight adjustment to the shape of the lambda parameters. Maybe instead of `(curr_1, next_1), (curr_2, next_2), ...` it's easier to pass `(curr_1, curr_2, ...), (next_1, next_2, ...)`.